File: sudoku_cpp/sudoku.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stdexcept>
#include <sstream>
#include <iterator>
#include <algorithm>
#include <set>
#include <map>
#include <utility>

class Grid {
public:
    static const int BOX_SIZE = 3;
    static const int GRID_SIZE = 9;

    Grid(const std::string& grid_string) {
        grid = parse_grid(grid_string);
        validate_grid();
    }

    // 构造函数用于反序列化
    Grid(const std::vector<std::vector<int>>& grid_) : grid(grid_) {
        validate_grid();
    }

    // 克隆构造函数
    Grid(const Grid& other) : grid(other.grid) {}

    // 获取指定行
    std::vector<int> getRow(int row) const {
        if (!(0 <= row && row < GRID_SIZE)) {
            throw std::out_of_range("行索引超出范围。");
        }
        return grid[row];
    }

    // 获取指定列
    std::vector<int> getColumn(int col) const {
        if (!(0 <= col && col < GRID_SIZE)) {
            throw std::out_of_range("列索引超出范围。");
        }
        std::vector<int> column(GRID_SIZE);
        for (int row = 0; row < GRID_SIZE; ++row) {
            column[row] = grid[row][col];
        }
        return column;
    }

    // 获取指定位置的3x3方格
    std::vector<int> getBox(int row, int col) const {
        if (!(0 <= row && row < GRID_SIZE && 0 <= col && col < GRID_SIZE)) {
            throw std::out_of_range("行或列索引超出范围。");
        }
        int box_row = (row / BOX_SIZE) * BOX_SIZE;
        int box_col = (col / BOX_SIZE) * BOX_SIZE;
        std::vector<int> box;
        for (int r = box_row; r < box_row + BOX_SIZE; ++r) {
            for (int c = box_col; c < box_col + BOX_SIZE; ++c) {
                box.push_back(grid[r][c]);
            }
        }
        return box;
    }

    // 克隆方法
    Grid clone() const {
        return Grid(*this);
    }

    // 序列化
    std::string serialize() const {
        std::ostringstream oss;
        for (const auto& row : grid) {
            for (const auto& num : row) {
                oss << num;
            }
        }
        return oss.str();
    }

    // 反序列化
    static Grid deserialize(const std::string& grid_string) {
        return Grid(grid_string);
    }

    // 重载输出运算符
    friend std::ostream& operator<<(std::ostream& os, const Grid& grid) {
        for (const auto& row : grid.grid) {
            for (const auto& num : row) {
                os << num << ' ';
            }
            os << '\n';
        }
        return os;
    }

    // 重载相等运算符
    bool operator==(const Grid& other) const {
        return grid == other.grid;
    }

protected:
    std::vector<std::vector<int>> grid;

    // 解析输入字符串
    std::vector<std::vector<int>> parse_grid(const std::string& grid_string) {
        if (grid_string.length() != GRID_SIZE * GRID_SIZE) {
            throw std::invalid_argument("输入字符串长度必须为81位。");
        }
        if (!std::all_of(grid_string.begin(), grid_string.end(), ::isdigit)) {
            throw std::invalid_argument("输入字符串必须仅包含数字。");
        }

        std::vector<std::vector<int>> grid(GRID_SIZE, std::vector<int>(GRID_SIZE));
        for (int i = 0; i < GRID_SIZE; ++i) {
            for (int j = 0; j < GRID_SIZE; ++j) {
                char c = grid_string[i * GRID_SIZE + j];
                grid[i][j] = c - '0';
            }
        }
        return grid;
    }

    // 验证棋盘是否合法
    void validate_grid() {
        // 检查行
        for (int i = 0; i < GRID_SIZE; ++i) {
            std::vector<int> row = getRow(i);
            std::vector<int> nums;
            for (int num : row) {
                if (num != 0) nums.push_back(num);
            }
            std::set<int> num_set(nums.begin(), nums.end());
            if (nums.size() != num_set.size()) {
                throw std::invalid_argument("第" + std::to_string(i + 1) + "行存在重复的数字。");
            }
        }
        // 检查列
        for (int i = 0; i < GRID_SIZE; ++i) {
            std::vector<int> col = getColumn(i);
            std::vector<int> nums;
            for (int num : col) {
                if (num != 0) nums.push_back(num);
            }
            std::set<int> num_set(nums.begin(), nums.end());
            if (nums.size() != num_set.size()) {
                throw std::invalid_argument("第" + std::to_string(i + 1) + "列存在重复的数字。");
            }
        }
        // 检查3x3方格
        for (int box_row = 0; box_row < GRID_SIZE; box_row += BOX_SIZE) {
            for (int box_col = 0; box_col < GRID_SIZE; box_col += BOX_SIZE) {
                std::vector<int> box = getBox(box_row, box_col);
                std::vector<int> nums;
                for (int num : box) {
                    if (num != 0) nums.push_back(num);
                }
                std::set<int> num_set(nums.begin(), nums.end());
                if (nums.size() != num_set.size()) {
                    throw std::invalid_argument("位置 (" + std::to_string(box_row + 1) + ", " + std::to_string(box_col + 1) + ") 的3x3方格存在重复的数字。");
                }
            }
        }
    }
};

class Sudoku : public Grid {
public:
    Sudoku(const std::string& grid_string) : Grid(grid_string) {}

    Sudoku(const std::vector<std::vector<int>>& grid_) : Grid(grid_) {}

    Sudoku(const Grid& grid_) : Grid(grid_) {}

    // 获取推理结果（候选值）
    std::map<std::pair<int, int>, std::vector<int>> getInference() const {
        std::map<std::pair<int, int>, std::vector<int>> candidates;
        for (int row = 0; row < GRID_SIZE; ++row) {
            for (int col = 0; col < GRID_SIZE; ++col) {
                if (grid[row][col] == 0) {
                    std::vector<int> cell_candidates = get_candidates(row, col);
                    if (cell_candidates.empty()) {
                        throw std::invalid_argument("单元格 (" + std::to_string(row + 1) + ", " + std::to_string(col + 1) + ") 无可用候选值，数独无解。");
                    }
                    candidates[std::make_pair(row, col)] = cell_candidates;
                }
            }
        }
        return candidates;
    }

    // 获取指定单元格的候选值
    std::vector<int> get_candidates(int row, int col) const {
        std::set<int> used_vals;
        std::vector<int> row_vals = getRow(row);
        std::vector<int> col_vals = getColumn(col);
        std::vector<int> box_vals = getBox(row, col);
        for (int val : row_vals) {
            if (val != 0) used_vals.insert(val);
        }
        for (int val : col_vals) {
            if (val != 0) used_vals.insert(val);
        }
        for (int val : box_vals) {
            if (val != 0) used_vals.insert(val);
        }
        std::vector<int> candidates;
        for (int n = 1; n <= 9; ++n) {
            if (used_vals.find(n) == used_vals.end()) {
                candidates.push_back(n);
            }
        }
        return candidates;
    }
};

// 测试代码
void test_sudoku() {
    std::string input_string = "017903600000080000900000507072010430000402070064370250701000065000030000005601720";
    Sudoku sudoku(input_string);

    std::cout << "数独棋盘：" << std::endl;
    std::cout << sudoku << std::endl;

    std::map<std::pair<int, int>, std::vector<int>> inference;
    try {
        inference = sudoku.getInference();
    } catch (const std::exception& e) {
        std::cerr << "错误：" << e.what() << std::endl;
        return;
    }

    std::cout << "\n推理结果（候选值）：" << std::endl;
    for (std::map<std::pair<int, int>, std::vector<int>>::const_iterator it = inference.begin(); it != inference.end(); ++it) {
        const std::pair<int, int>& key = it->first;
        const std::vector<int>& value = it->second;
        std::cout << "单元格 (" << key.first + 1 << ", " << key.second + 1 << ") 的候选值: ";
        for (int val : value) {
            std::cout << val << ' ';
        }
        std::cout << std::endl;
    }

    // 克隆测试
    Sudoku cloned_sudoku = sudoku.clone();
    std::cout << "\n克隆后的数独棋盘：" << std::endl;
    std::cout << cloned_sudoku << std::endl;

    // 序列化和反序列化测试
    std::string serialized = sudoku.serialize();
    std::cout << "\n序列化的数独棋盘：" << std::endl;
    std::cout << serialized << std::endl;

    Sudoku deserialized_sudoku = Sudoku::deserialize(serialized);
    std::cout << "\n反序列化后的数独棋盘：" << std::endl;
    std::cout << deserialized_sudoku << std::endl;

    // 比较测试
    std::cout << "\n比较两个Sudoku对象是否相等：";
    std::cout << (sudoku == cloned_sudoku ? "相等" : "不相等") << std::endl;
}

int main() {
    test_sudoku();
    return 0;
}

File: sudoku_python/sudoku.py
import copy
import json

class Grid:
    BOX_SIZE = 3
    GRID_SIZE = 9
    def __init__(self, grid_string):
        self.grid = self.parse_grid(grid_string)
        self.validate_grid()

    def parse_grid(self, grid_string):
        if len(grid_string) != self.GRID_SIZE * self.GRID_SIZE:
            raise ValueError("输入字符串长度必须为81位。")
        if not grid_string.isdigit():
            raise ValueError("输入字符串必须仅包含数字。")
        grid = [[int(grid_string[i * self.GRID_SIZE + j]) for j in range(self.GRID_SIZE)] for i in
                range(self.GRID_SIZE)]
        return grid

    def validate_grid(self):
        """验证初始棋盘是否符合数独规则。"""
        for i in range(self.GRID_SIZE):
            row = [num for num in self.getRow(i) if num != 0]
            if len(row) != len(set(row)):
                raise ValueError(f"第{i + 1}行存在重复的数字。")

            col = [num for num in self.getColumn(i) if num != 0]
            if len(col) != len(set(col)):
                raise ValueError(f"第{i + 1}列存在重复的数字。")

        for box_row in range(0, self.GRID_SIZE, self.BOX_SIZE):
            for box_col in range(0, self.GRID_SIZE, self.BOX_SIZE):
                box = [num for num in self.getBox(box_row, box_col) if num != 0]
                if len(box) != len(set(box)):
                    raise ValueError(f"位置 ({box_row + 1}, {box_col + 1}) 的3x3方格存在重复的数字。")

    def getRow(self, row):
        """返回指定行的数据"""
        if not 0 <= row < self.GRID_SIZE:
            raise IndexError("行索引超出范围。")
        return self.grid[row]

    def getColumn(self, col):
        """返回指定列的数据"""
        if not 0 <= col < self.GRID_SIZE:
            raise IndexError("列索引超出范围。")
        return [self.grid[row][col] for row in range(self.GRID_SIZE)]

    def getBox(self, row, col):
        """返回指定位置的3x3方格数据"""
        if not (0 <= row < self.GRID_SIZE and 0 <= col < self.GRID_SIZE):
            raise IndexError(f"行或列索引超出范围：row={row}, col={col}")
        box_row = (row // self.BOX_SIZE) * self.BOX_SIZE
        box_col = (col // self.BOX_SIZE) * self.BOX_SIZE
        return [self.grid[r][c] for r in range(box_row, box_row + self.BOX_SIZE)
                for c in range(box_col, box_col + self.BOX_SIZE)]

    def __str__(self):
        """外表化：输出友好的字符串表示形式"""
        return '\n'.join([' '.join(map(str, row)) for row in self.grid])

    def __eq__(self, other):
        """比较两个Grid对象是否相等"""
        if isinstance(other, Grid):
            return self.grid == other.grid
        return False

    def __hash__(self):
        """生成哈希值"""
        return hash(tuple(tuple(row) for row in self.grid))

    def clone(self):
        """克隆：返回一个Grid对象的深度副本"""
        return copy.deepcopy(self)

    def serialize(self):
        """串行化：将Grid对象转换为JSON格式"""
        return json.dumps(self.grid)

    @classmethod
    def deserialize(cls, json_string):
        """反序列化：从JSON字符串创建一个Grid对象"""
        grid_list = json.loads(json_string)
        grid_string = ''.join(str(num) for row in grid_list for num in row)
        return cls(grid_string)

    def __iter__(self):
        return iter(self.grid)


class Sudoku(Grid):
    def __init__(self, grid_string=None, grid=None):
        if grid is not None:
            self.grid = grid
            self.validate_grid()
        elif grid_string is not None:
            super().__init__(grid_string)
        else:
            raise ValueError("必须提供grid_string或grid之一。")


    def getInference(self):
        """推理棋盘，返回每个单元格的候选值"""
        candidates = {}
        for row in range(self.GRID_SIZE):
            for col in range(self.GRID_SIZE):
                if self.grid[row][col] == 0:
                    cell_candidates = self.get_candidates(row, col)
                    if not cell_candidates:
                        raise ValueError(f"单元格 ({row + 1}, {col + 1}) 无可用候选值，数独无解。")
                    candidates[(row, col)] = cell_candidates
        return candidates

    def get_candidates(self, row, col):
        """返回指定单元格的候选值"""
        row_vals = set(self.getRow(row))
        col_vals = set(self.getColumn(col))
        box_vals = set(self.getBox(row, col))
        used_vals = row_vals | col_vals | box_vals
        return [n for n in range(1, 10) if n not in used_vals]

    def __str__(self):
        """外表化：输出数独棋盘的字符串表示"""
        return super().__str__()

    def __eq__(self, other):
        """比较两个Sudoku对象是否相等"""
        return super().__eq__(other)

    def clone(self):
        """克隆：返回一个Sudoku对象的深度副本"""
        return copy.deepcopy(self)

    def serialize(self):
        """串行化：将Sudoku对象转换为JSON格式"""
        return super().serialize()

    @classmethod
    def deserialize(cls, json_string):
        grid = json.loads(json_string)
        return cls(grid=grid)


# 测试代码
def test_sudoku():
    input_string = "017903600000080000900000507072010430000402070064370250701000065000030000005601720"
    sudoku = Sudoku(input_string)

    print("数独棋盘：")
    print(sudoku)

    inference = sudoku.getInference()
    print("\n推理结果（候选值）：")
    for key, value in inference.items():
        print(f"单元格 {key} 的候选值: {value}")

    # 克隆测试
    cloned_sudoku = sudoku.clone()
    print("\n克隆后的数独棋盘：")
    print(cloned_sudoku)

    # 序列化和反序列化测试
    serialized = sudoku.serialize()
    print("\n序列化的数独棋盘：")
    print(serialized)

    deserialized_sudoku = Sudoku.deserialize(serialized)
    print("\n反序列化后的数独棋盘：")
    print(deserialized_sudoku)

    input_string = "017903600000080000900000507072010430000402070064370250701000065000030000005601720"
    sudoku2 = Sudoku(input_string)

    # 比较测试
    print("\n比较两个Sudoku对象是否相等：", sudoku == sudoku2)


if __name__ == "__main__":
# 调用测试代码
    test_sudoku()
